File: Cpp/Problem_1496.cpp
#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

/*
Problem - 1496: https://leetcode.com/problems/path-crossing/
C++ Solution!
*/
class Solution {
public:
	struct hashFunction {
		size_t operator()(const pair<int, int> &x) const {
			return x.first ^ x.second;
		}
	};

	struct pair_hash {
		size_t operator()(const pair<int, int> &v) const {
			return v.first * 31 + v.second;
		}
	};

	bool isPathCrossing(string path) {
		int x_pos = 0, y_pos = 0;
		unordered_set<pair<int, int>, pair_hash> prev_pos;
		prev_pos.insert({ 0, 0 });

		for (char i : path) {
			switch (i)
			{
			case 'N':
				x_pos++;
				break;
			case 'S':
				x_pos--;
				break;
			case 'E':
				y_pos++;
				break;
			case 'W':
				y_pos--;
				break;
			default:
				break;
			}

			if (prev_pos.find({ x_pos, y_pos }) != prev_pos.end())
				return true;
			else
				prev_pos.insert({ x_pos, y_pos });
		}

		return false;
	}
};

int main() {
	Solution sol;

	cout << sol.isPathCrossing("NES") << endl;
	cout << sol.isPathCrossing("NESWW") << endl;

	return 0;
}

File: Python/Problem_1496.py
# Problem - 1496: https://leetcode.com/problems/path-crossing/
# Python3 Solution!
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        prev_pos = set()
        x, y = 0, 0

        prev_pos.add((x, y))

        for direction in path:
            if direction == 'N':
                x+=1
            elif direction == 'S':
                x-=1
            elif direction == 'E':
                y+=1
            else:
                y-=1

            if (x , y) in prev_pos:
                return True
            else:
                prev_pos.add((x, y))

        return False


if __name__=="__main__":
    sol = Solution()

    print(sol.isPathCrossing("NES"))
    print(sol.isPathCrossing("NESWW"))